// rosreact/dist/esm/components/RosConnection/RosConnection.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { sha512 } from 'js-sha512';
import PropTypes from 'prop-types';
import React, { createContext, useContext, useEffect, useRef, useState, } from 'react';
import { Ros } from 'roslib';
import { getRosObject } from './RosInstanceManager';
var RosContext = createContext(new Ros({}));
var DefaultRosProps = {
    url: 'ws://127.0.0.1:9090',
    autoConnect: false,
    autoConnectTimeout: 1000,
    authenticate: false,
    user: '',
    password: '',
};
export var RosConnection = function (_a) {
    var children = _a.children, userProps = __rest(_a, ["children"]);
    var props = __assign(__assign({}, DefaultRosProps), userProps);
    var _b = useState(props.url), url = _b[0], setUrl = _b[1];
    var rosRef = useRef(new Ros({}));
    // Trigger update if URL changes
    useEffect(function () {
        setUrl(props.url);
    }, [props.url]);
    useEffect(function () {
        rosRef.current = getRosObject(url);
        setupConnectionCallbacks(rosRef.current, url, props.autoConnect, props.autoConnectTimeout, props.authenticate, props.user, props.password);
        connect(rosRef.current, url, props.authenticate, props.user, props.password);
        return function () {
            closeConnection(rosRef.current);
        };
    }, [
        url,
        props.authenticate,
        props.autoConnect,
        props.autoConnectTimeout,
        props.password,
        props.user,
    ]);
    return React.createElement(RosContext.Provider, { value: rosRef.current }, children);
};
RosConnection.propTypes = {
    children: PropTypes.node.isRequired,
    url: PropTypes.string.isRequired,
    autoConnect: PropTypes.bool,
    autoConnectTimeout: PropTypes.number,
    authenticate: PropTypes.bool,
    user: PropTypes.string,
    password: PropTypes.string,
};
export function setupConnectionCallbacks(ros, url, autoConnect, autoConnectTimeout, authenticate, user, password) {
    if (url === void 0) { url = DefaultRosProps.url; }
    if (autoConnect === void 0) { autoConnect = DefaultRosProps.autoConnect; }
    if (autoConnectTimeout === void 0) { autoConnectTimeout = DefaultRosProps.autoConnectTimeout; }
    if (authenticate === void 0) { authenticate = DefaultRosProps.authenticate; }
    if (user === void 0) { user = DefaultRosProps.user; }
    if (password === void 0) { password = DefaultRosProps.password; }
    ros.on('connection', function () {
        console.log('Connected');
    });
    ros.on('close', function () {
        console.log('Disconnected');
    });
    ros.on('error', function () {
        console.log('Connection error');
        // Attempt to reconnect
        if (autoConnect) {
            setTimeout(function () {
                connect(ros, url, authenticate, user, password);
            }, autoConnectTimeout);
        }
    });
}
export function connect(ros, url, authenticate, user, password) {
    if (url === void 0) { url = DefaultRosProps.url; }
    if (authenticate === void 0) { authenticate = DefaultRosProps.authenticate; }
    if (user === void 0) { user = DefaultRosProps.user; }
    if (password === void 0) { password = DefaultRosProps.password; }
    ros.connect(url);
    if (authenticate) {
        var authMessage = new AuthenticationMessage(url, user, password);
        ros.authenticate(authMessage.getMac(), authMessage.client, authMessage.dest, authMessage.rand, authMessage.time, authMessage.level, authMessage.timeEnd);
    }
}
export function closeConnection(ros) {
    ros.close();
}
var AuthenticationMessage = /** @class */ (function () {
    function AuthenticationMessage(url, user, password) {
        this.dest = url;
        this.client = user;
        this.secret = password;
        this.rand = 'randomstring';
        this.time = new Date().getTime();
        this.level = 'user';
        this.timeEnd = this.time;
    }
    AuthenticationMessage.prototype.getMac = function () {
        return sha512(this.secret +
            this.client +
            this.dest +
            this.rand +
            this.time.toString() +
            this.level +
            this.timeEnd.toString());
    };
    return AuthenticationMessage;
}());
export function useRos() {
    var ros = useContext(RosContext);
    if (ros === undefined) {
        throw new Error('rosreact components must be wrapped by a RosProvider');
    }
    return ros;
}
