// rosreact/dist/cjs/components/Subscriber/useSubscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSubscription = void 0;
var react_1 = require("react");
var RosConnection_1 = require("../RosConnection");
var getTopic_1 = require("./getTopic");
function useSubscription(props) {
    var _a;
    var ros = (0, RosConnection_1.useRos)();
    var topic = props.topic, messageType = props.messageType, throttleRate = props.throttleRate, latch = props.latch, queueLength = props.queueLength, queueSize = props.queueSize, customCallback = props.customCallback, compareFunc = props.compareFunc;
    var _b = (0, react_1.useState)((_a = props.messageInitialValue) !== null && _a !== void 0 ? _a : null), message = _b[0], setMessage = _b[1];
    var messageRef = (0, react_1.useRef)(message);
    (0, react_1.useEffect)(function () {
        var updateMessage = function (msg) {
            setMessage(msg);
            messageRef.current = msg;
        };
        var messageCallback = function (newMsg) {
            if (compareFunc) {
                var compareRes = compareFunc(messageRef.current, newMsg);
                // The values are the same, return.
                if (compareRes === true || compareRes === 0) {
                    return;
                }
            }
            if (customCallback) {
                customCallback(newMsg);
            }
            updateMessage(newMsg);
        };
        var subscription = (0, getTopic_1.subscribe)(ros, {
            topic: topic,
            messageType: messageType,
            throttleRate: throttleRate,
            latch: latch,
            queueLength: queueLength,
            queueSize: queueSize,
        }, messageCallback);
        return function () {
            (0, getTopic_1.unsubscribe)(subscription, messageCallback);
        };
    }, [
        ros,
        topic,
        messageType,
        throttleRate,
        latch,
        queueLength,
        queueSize,
        customCallback,
        compareFunc,
    ]);
    return message;
}
exports.useSubscription = useSubscription;
